src/prometheus/variant_counter.zig

```diff
@@ -0,0 +1,143 @@
+const std = @import("std");
+const prometheus = @import("lib.zig");
+
+const Atomic = std.atomic.Value;
+
+const Metric = prometheus.metric.Metric;
+const MetricType = prometheus.metric.MetricType;
+
+/// Separately count the occurrence of each variant within an enum or error set.
+///
+/// This is a composite metric, similar to a histogram, except the "buckets" are
+/// discrete, unordered, and identified by name, instead of representing numeric ranges.
+pub fn VariantCounter(comptime EnumOrError: type) type {
+    const indexer = VariantIndexer.init(EnumOrError);
+    const names = indexer.names();
+
+    return struct {
+        counts: [names.len]Atomic(u64) = .{Atomic(u64).init(0)} ** names.len,
+        metric: Metric = .{ .getResultFn = getResult },
+
+        const Self = @This();
+
+        pub const metric_type: MetricType = .variant_counter;
+        pub const Type = EnumOrError;
+
+        pub fn observe(self: *Self, err: EnumOrError) void {
+            _ = self.counts[indexer.index(err)].fetchAdd(1, .monotonic);
+        }
+
+        pub fn getResult(metric: *Metric, _: std.mem.Allocator) Metric.Error!Metric.Result {
+            const self: *Self = @fieldParentPtr("metric", metric);
+            return .{ .variant_counter = .{
+                .counts = &self.counts,
+                .names = &names,
+            } };
+        }
+    };
+}
+
+pub const VariantCounts = struct {
+    counts: []const Atomic(u64),
+    names: []const []const u8,
+};
+
+/// Assigns a continuous sequence of integers starting at 0 to the
+/// variants. For example, if you have an variant set with 10 variants,
+/// it will assign the numbers 0-9 to each of the variants.
+///
+/// Enums usually already have numbers with this property, but it's
+/// not guaranteed. This is particularly useful for variants, whose
+/// integers are likely not continuous.
+///
+/// You can provide an variant to determine its index, or provide an
+/// index to get the respective variant.
+///
+/// This struct must be initialized at comptime, but it can be used
+/// at any time.
+const VariantIndexer = struct {
+    EnumOrError: type,
+    offset: u16,
+    index_to_int: []const u16,
+    int_to_index: []const u16,
```

Comment:
not sure why we need all this -- `toInt` `fromInt` and `toName` can just use `@intFromEnum/fromError` at runtime (rn its like a cache for a really cheap operation?)

Reply:
There's an array of counters for each error. You need to be able to associate error to one of those counters. Let's say you have three errors, then you'd like to assign each of those errors one of the numbers 0,1,2 so you can identify which counter they are associated with in the array of counters. That's the job of this struct.

The three errors that you're tracking may have integer representations like 121, 249, and 310. They need to be assumed that they are basically random. The goal here is to map those numbers to 0, 1, and 2.

Reply:
huh ok that makes sense i was unaware that error sets would map to random integer representations (assumed it would be incremental from 0)

Reply:
I put something like that in the struct's doc, but I accidentally deleted the word "error". It was supposed to say this:
```
/// Assigns a continuous sequence of integers starting at 0 to the
/// variants. For example, if you have an variant set with 10 variants,
/// it will assign the numbers 0-9 to each of the variants.
///
/// Enums usually already have numbers with this property, but it's
/// not guaranteed. This is particularly useful for **errors**, whose
/// integers are likely not continuous.
```
Note where it says **errors** (currently in the code it says "variants" which is a mistake)

Is this good enough or do you think it needs further explanation?

Reply:
Also FYI, error integers are globally unique. They're not only unique within an error set. This is required because error sets need to intersect with each other. The same error needs to be usable across multiple error sets. The most extreme example is `anyerror`, which is the set of all errors. Enums on the other hand, are strictly non-intersecting sets. But then you have the option to specify arbitrary integers for each variant, so it could also be non-continuous.

It's unspecified how error integers are assigned by the compiler. I've seen large gaps between the integers for errors within the same set. They're also not guaranteed to be consistent across different compilations.

Reply:
yeah thats good -- huh very interesting thanks for explaining